src/azure_devtools/scenario_tests/preparers.py

```diff
@@ -40,8 +40,12 @@ def _preparer_wrapper(test_class_instance, **kwargs):
                     resource_name,
                     **kwargs
                 )
-            test_class_instance.addCleanup(
-                lambda: self.remove_resource_with_record_override(resource_name, **kwargs)
+
+            # Insert the cleanup function at the beginning to enforce Russian Doll effect - the last declared resource
+            # to be deleted first.
+            test_class_instance._cleanups.insert(  # pylint: disable=protected-access
```

Comment:
I don't like that at all, we don't know how this is used somewhere. Russian roulette :p 

I would rather use a more teardown approach without using `addCleanup` at all
```python
try:
    fn(test_class_instance, **kwargs)
except Exception:
    raise
finally:
    self.remove_resource_with_record_override(resource_name, **kwargs)
```

Reply:
What's the downside of using `addCleanup` in your opinion?

Reply:
First, you don't control what is inside `cleanUp`, since this is a public API for the test guy. Maybe you just mess up his order by doing that.
Second, I feel in this particular example, your recursively put together decorator, and then cleaning should be done the same way. And that's simple to do simple without touching a private member. 
Do you have any concerns with the `try`/`finally` approach?